Enhance wxDynamicLibrary error logging and simplify code

lanurmi opened the pull request:
Enhancements to logging, error handling, and more.

lanurmi committed:
Use the same function for logging dynlib errors on all platforms

Reduces the amount of #ifdefs scattered all over the code, thereby
simplifying the code.

The function was renamed from Error() to ReportError() to emphasize what
its purpose is.

Error messages logged on *nix are now a bit more verbose, as they are
prefixed with our own description text, which were earlier omitted on
platforms using the dlxxx() API.

vadz commented on include/wx/dynlib.h:
```diff
@@ -251,10 +251,18 @@ class WXDLLIMPEXP_BASE wxDynamicLibrary
 
     // unload the library, also done automatically in dtor
-    void Unload() { if ( IsLoaded() ) { Unload(m_handle); m_handle = NULL; } }
+    bool Unload() {
```
```suggestion
    bool Unload()
    {
```

vadz commented on include/wx/dynlib.h:
```diff
@@ -357,11 +365,8 @@ class WXDLLIMPEXP_BASE wxDynamicLibrary
-#endif // HAVE_DLERROR
-
+    // log the error after an OS dynamic library function failure
+    static void ReportError(const wxString &msg, const wxString &name);
```
```suggestion
    static void ReportError(const wxString& msg, const wxString& name);
```

vadz commented on interface/wx/dynlib.h:
```diff
@@ -232,8 +232,10 @@ class wxDynamicLibrary
         calls this method from its destructor if it had been successfully
         loaded.
+
+        Returns @true if the library was successfully unloaded, @false otherwise.
```
```suggestion
        Returns @true if the library was successfully unloaded or wasn't loaded in the first place, @false otherwise.
```

vadz commented on src/msw/dlmsw.cpp:
```diff
@@ -145,6 +145,28 @@ wxDllType wxDynamicLibrary::GetProgramHandle()
+// ----------------------------------------------------------------------------
+
+/* static */
+void wxDynamicLibrary::ReportError(const wxString &msg, const wxString &name)
```
```suggestion
void wxDynamicLibrary::ReportError(const wxString& msg, const wxString& name)
```

vadz commented on interface/wx/dynlib.h:
```diff
@@ -243,8 +245,10 @@ class wxDynamicLibrary
         wxDynamicLibrary object. In this case you may call Detach() and store
         the handle somewhere and call this static method later to unload it.
+
+        Returns @true if the library was successfully unloaded, @false otherwise.
```
Same as above.

vadz commented on src/unix/dlunix.cpp:
```diff
@@ -106,22 +106,21 @@ wxDllType wxDynamicLibrary::RawLoad(const wxString& libname, int flags)
     if ( rc != 0 )
-        Error();
-#endif
+        ReportError(_("Failed to unload shared library"),
```
Won't this fail the assert you've added?

lanurmi replied:
It will, and it shouldn't. I'll fix it asap.

vadz reviewed (commented):
Thanks, I agree that simplification part is worth doing, but I think there is a real problem in the Unix version, please see below. And if we do want to allow calling it with an empty second argument, we should probably give it the default `= wxString()` value.

I also think we could further simplify the code by dropping `HAVE_DLERROR` entirely: I'm pretty sure there are no platforms without it still in use.

Finally, I'm not really sure if it's worth complicating things by returning `bool` from `Unload()`: first of all, I've never actually seen `FreeLibrary/dlclose` fail (hang/deadlock/crash -- yes, but return with an error -- no). Second, even if it does fail, what are you supposed to do about it, anyhow?

lanurmi commented:
> Thanks, I agree that simplification part is worth doing, but I think there is a real problem in the Unix version, please see below. And if we do want to allow calling it with an empty second argument, we should probably give it the default `= wxString()` value.

That would make sense, yes.

> I also think we could further simplify the code by dropping `HAVE_DLERROR` entirely: I'm pretty sure there are no platforms without it still in use.

Indeed, everything about `shl_xxx()` points to HP-UX, and one source suggests even HP-UX has had `dlxxx()` functions since at least 19 years, probably more.

> Finally, I'm not really sure if it's worth complicating things by returning `bool` from `Unload()`: first of all, I've never actually seen `FreeLibrary/dlclose` fail (hang/deadlock/crash -- yes, but return with an error -- no). Second, even if it does fail, what are you supposed to do about it, anyhow?

Admittedly I haven't seen any of those fail either, except in the case that the library wasn't loaded to begin with. Also I don't have a real-world example about doing something in case of failure. But still, all the three currently supported native APIs do return a success/failure status for unloading a library, yet the status is explicitly suppressed by wx. If it was like 50% of the APIs have no way of obtaining the status, then it would be reasonable not to return it for the other half either, but here we have a rare case of all native APIs implementing the same functionality. However, I can de-cherry-pick the last commit from this PR, if desired.

vadz commented:
Thanks for the updates!

There is still one comment about the interface comment remaining, could you please make it consistent with the other one too?

There is also one test for `HAVE_DLERROR` which remains, I'm not sure if this is intentional? I think we should get rid of it too and also of the check for it in `configure.in`. We probably could also get rid of `dlopen()` check too, but this is a different matter...

Finally, concerning the error code from `Unload()`: I really think it's not useful. In the future we ought to mark functions returning error codes with `[[nodiscard]]` and forcing people to check the return result when they can't do anything about it just doesn't seem useful. Then there is also the ambiguity of the return value in case the library isn't loaded (both variants could make sense to me here). It seems simpler to just not bother with it.

lanurmi committed:
Get rid of non-POSIX code for loading dynlibs on *nix

The alternative, (non-POSIX) shl_xxx() API is/was apparently available on
HP-UX, but even there the POSIX dlxxx() functions have been the preferred
way to load libraries since the past ~20 years.

lanurmi commented:
> There is still one comment about the interface comment remaining, could you please make it consistent with the other one too?
> 
> There is also one test for `HAVE_DLERROR` which remains, I'm not sure if this is intentional?
> 
> Finally, concerning the error code from `Unload()`: I really think it's not useful.

My latest changes are now squashed and pushed. I removed the `Unload()` error code commit from this PR. Further `#ifdef`s along with the code calling `shl_xxx()` API is removed in a separate commit.

vadz committed:
Remove unnecessary c_str() call

No real changes.

vadz committed:
Remove checks for shl_load() from configure/CMake

This function is not used any more since e289eb07e1 (Get rid of
non-POSIX code for loading dynlibs on *nix, 2020-05-13), so don't check
for it and don't defined the corresponding HAVE_SHL_LOAD symbol.

vadz committed:
Assume dlerror() function is always available when dlopen() is

Remove separate checks for dlerror() which don't seem to be needed under
any platform any longer.

No real changes, just slim down configure/CMake a tiny bit.

vadz commented:
Thanks, I've removed the corresponding checks from configure/CMake too, and will merge it when the CI builds pass.
